### DISCUSSION TITLE
XLocalizer in a console app without services.AddRazorPages

### DISCUSSION THREAD

**Discussion by julianadormon** [LazZiya/XLocalizer#49] [General]
@LazZiya I don't mean to bother you so much, but I have run into an issue using XLocalizer in my console app.

Previously you helped me get it working in my console application which generates receipts in a loop.
I used the CultureSwitcher to switch cultures in the loop depending on the customer's culture who is receiving the looped receipt.

I was generating PDFs using a library that fetches an external webpage and converts the HTML to PDF. I have since switched to generating the view using Microsoft.Playwrite which renders the HTML locally without having to go to the internet for it.

Unfortunately, Playwrite does not seem to work well when services.AddRazorPages() is used, and this means I cannot hook up services.AddRazorPages().AddXLocalizer<>().

With the help pf chatGPT, I was able to wire up IStringLocalizer as follows

`// Configure Request Localization Options
            services.Configure<RequestLocalizationOptions>(options =>
            {
                var SupportedCultures = new List<CultureInfo>()
                {
                    new CultureInfo("en-US"),
                    new CultureInfo("fr"),
                    new CultureInfo("es")
                };
                options.DefaultRequestCulture = new RequestCulture("en-US");
                options.SupportedCultures = SupportedCultures;
                options.SupportedUICultures = SupportedCultures;
            });
            
            // Configure caching for XLocalizer
            services.AddSingleton<IMemoryCache, MemoryCache>();
            services.TryAddSingleton<ExpressMemoryCache>();

            // Configure Options for XLocalizer
            services.Configure<XLocalizerOptions>(options =>
            {
                options.ResourcesPath = "LocalizationResources";
                options.AutoAddKeys = true;
                options.AutoTranslate = true;
            });
            
            // Register GoogleTranslateService as the default translator and TranslatorFactory
            services.AddHttpClient<GoogleTranslateService>();
            services.AddSingleton<ITranslatorFactory, TranslatorFactory<GoogleTranslateService>>();

            // Manually configure XmlResourceProvider with dependencies
            services.AddSingleton<IXResourceProvider, XmlResourceProvider>(sp =>
            {
                var Options = sp.GetRequiredService<IOptions<XLocalizerOptions>>();
                var Logger = sp.GetRequiredService<ILogger<XmlResourceProvider>>();
                return new XmlResourceProvider(Options, Logger);
            });

            // Register IStringLocalizer<LocSource> manually with all dependencies
            services.AddSingleton<IStringLocalizer<LocSource>>(sp =>
            {
                var Cache = sp.GetRequiredService<ExpressMemoryCache>();
                var TranslatorFactory = sp.GetRequiredService<ITranslatorFactory>();
                var ResourceProvider = sp.GetRequiredService<IXResourceProvider>();
                var LocalizerOptions = sp.GetRequiredService<IOptions<XLocalizerOptions>>();
                var LocalizationOptions = sp.GetRequiredService<IOptions<RequestLocalizationOptions>>();
                var LoggerFactory = sp.GetRequiredService<ILoggerFactory>();

                return new XStringLocalizer<LocSource>(
                    Cache,
                    TranslatorFactory,
                    ResourceProvider,
                    LocalizerOptions,
                    LocalizationOptions,
                    LoggerFactory);
            });`

While this allows the program to run, if I set any culture other than the default culture, I get the following error:

System.NullReferenceException: Object reference not set to an instance of an object.
   at XLocalizer.XStringLocalizer`1.GetLocalizedString(String name, Object[] arguments)
   at XLocalizer.XStringLocalizer`1.get_Item(String name)

Hoping you can help? I am happy to pay if you need to code something new for this :) Thanks in advance.

**Comment by julianadormon** [comment]
`using CultureSwitcher Cs = new("es");
                
                Logger.LogInformation($"Inside SendInvoicesAndReceipts > Thread.CultureCode: {Thread.CurrentThread.CurrentCulture.Name}");

                Console.WriteLine(Thread.CurrentThread.CurrentCulture.Name);
                
                var TestLocalizer = Localizer["Hello world"];`

**Comment by LazZiya** [comment]
Hi @julianadormon

Don't worry I'm not bothered, instead I feel glad to see different use cases of `XLocalizer` :)

Actually I didn't consider console apps support with `XLocalizer`, but it seems as a potential next step.

Re your issue, I see that you need to register some internal services to get it working. Have a look at the `.AddXLocalizer` extension method and try to register the internal services in a similar way:

https://github.com/LazZiya/XLocalizer/blob/07e1a002ffe54d2739bb43e46c9d2508f263864d/XLocalizer/DependencyInjection.cs#L65-L112

Maybe you can avoid adding data annotation and identity errors localization, sine they are not available in a console app.

Hopefully this will help, if not let me know please.

**Reply by julianadormon** [reply]
Thank you! I will try this.

**Comment by julianadormon** [comment]
Thank you @LazZiya ! That worked. Here is the code that worked for me, perhaps it can help someone else.

`
//Cannot use services.AddRazorPages with HtmlRenderer, so cannot inject XLocalizer, so going to instantiate this inside
            
            // Configure Request Localization Options
            services.Configure<RequestLocalizationOptions>(options =>
            {
                var SupportedCultures = new List<CultureInfo>()
                {
                    new CultureInfo("en-US"),
                    new CultureInfo("fr"),
                    new CultureInfo("es")
                };
                options.DefaultRequestCulture = new RequestCulture("en-US");
                options.SupportedCultures = SupportedCultures;
                options.SupportedUICultures = SupportedCultures;
            });
            
            services.AddHttpClient<ITranslator, GoogleTranslateService>();
            services.AddSingleton<IXResourceProvider, XmlResourceProvider>();

            // Configure Options for XLocalizer
            services.Configure<XLocalizerOptions>(options =>
            {
                options.ResourcesPath = "LocalizationResources"; // Path to the resource files
                options.AutoAddKeys = true; // Automatically add missing keys to resources
                options.AutoTranslate = true; // Automatically translate missing keys
            });

            // Configure caching for XLocalizer
            services.AddSingleton<IMemoryCache, MemoryCache>();
            services.TryAddSingleton<ExpressMemoryCache>();

            // // Register GoogleTranslateService as the default translator and TranslatorFactory
            // services.AddHttpClient<GoogleTranslateService>();
            services.AddSingleton<ITranslatorFactory, TranslatorFactory<GoogleTranslateService>>();
            
            services.TryAddSingleton<IXResourceProvider, ResxResourceProvider>();
            services.TryAddTransient<IXResourceExporter, XmlResourceExporter>();
            
            services.AddSingleton<IStringLocalizer, XStringLocalizer<LocSource>>();
            
            // Register generic IStringLocalizer<> and factories
            services.AddTransient(typeof(IStringLocalizer<>), typeof(XStringLocalizer<>));
            services.AddSingleton<IStringLocalizerFactory, XStringLocalizerFactory<LocSource>>();
            
            services.AddSingleton<IStringLocalizerFactory, XStringLocalizerFactory<LocSource>>();
            services.AddSingleton<IXStringLocalizerFactory, XStringLocalizerFactory<LocSource>>();`

**Reply by LazZiya** [reply]
Glad to know it worked 🚀 and thanks for sharing the working code ❤️
